### PR TITLE
JAMES-2413 [DLP] Mailet store emails into repositories depending on sender domains

### PR DESCRIPTION
**Pull request by trantienduchn**

### CODE REVIEW

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.james.mailrepository.api.MailRepositoryStore;
+import org.apache.mailet.Mail;
+import org.apache.mailet.base.GenericMailet;
+
+/**
+ * Stores incoming Mail in a repository defined by the sender's domain.<br>
+ *
+ * Supported configuration parameters:
+ *
+ *  - "urlPrefix" compulsory: defines the prefix for the per sender's domain repository.
```
Comment (rouazana) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java]: s/compulsory/mandatory/ ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ *  If equals to 'cassandra://var/mail/sendersRepositories/' a mail sent by 'user@domain.tld'
```
Comment (rouazana) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java]: s/If equals to/For example for the value/ ?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ *  will be stored in 'cassandra://var/mail/sendersRepositories/domain.tld'.
```
Comment (rouazana) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java]: do we need to have an option to create the repository if not already existing?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            .build();
+
+        mailet.service(mail);
+
+        assertThat(mail.getState())
+            .isEqualTo(Mail.GHOST);
+    }
+
+    @Test
+    void initShouldSetNotPassThroughWhenPassThroughtIsNotBooleans() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: s/Booleans/Set/

Comment (rouazana) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: s/Throught/Through

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            .sender(MailAddressFixture.SENDER)
+            .state(Mail.DEFAULT)
+            .build();
+
+        assertThatThrownBy(() -> mailet.service(mail))
+            .isInstanceOf(MessagingException.class);
+    }
+
+    @Test
+    void initShouldSetNotPassThroughWhenPassThroughtIsNotBoolean() throws Exception {
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: s/Throught/Through

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+        assertThat(mail.getState())
+            .isEqualTo(Mail.GHOST);
+    }
+
+    @Test
+    void getMailetInfoShouldReturnExpectedResult() {
+        assertThat(mailet.getMailetInfo())
+            .isEqualTo("ToSenderDomainRepository Mailet");
+    }
```
Comment (rouazana) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: can you at least add a test when the sender domain repository does not exist?

Comment (chibenwa) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: Not possible with `MockMailRepositoryStore` very sadly... You do have to provide the repository.

Mail repository store is bould to the injection framework which proposes two different implementation. So in unit tests we can not demonstrate this.

The solution might be to write a quick integration test. We will work on it tomorrow with Duc.

Comment (trantienduchn) [reply server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: The was a fault, the method name `initShouldSetNotPassThroughWhenPassThroughtIsNotBoolean` is duplicated, the test that @rouazana want may be described the method `initShouldSetNotPassThroughWhenPassThroughtIsNotBooleans`

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+import org.apache.mailet.base.GenericMailet;
+
+/**
+ * Stores incoming Mail in a repository defined by the sender's domain.<br>
+ *
+ * Supported configuration parameters:
+ *
+ *  - "urlPrefix" mandatory: defines the prefix for the per sender's domain repository.
+ *  For example for the value 'cassandra://var/mail/sendersRepositories/', a mail sent by 'user@domain.tld'
+ *  will be stored in 'cassandra://var/mail/sendersRepositories/domain.tld'.
```
Comment (aduprat) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java]: Can we have a domain name with a `/` in James ? ;)

Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java]: Should have no impact anyway: `/` is valid in a URL...

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java (hunk 3 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java]: For the record @rouazana  disabling repository creation is pointless if the admin can't create the mail repositories he wants. Hence I asked Duc to add https://github.com/linagora/james-project/pull/1457/commits/f0ee57225518c64e4a3f6e6e57b1efb76d51c5c5

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java [hunk 8]
```diff
@@ -37,21 +43,29 @@
  *  For example for the value 'cassandra://var/mail/sendersRepositories/', a mail sent by 'user@james.org'
  *  will be stored in 'cassandra://var/mail/sendersRepositories/james.org'.
  *  - "passThrough" optional, defaults to false. If true, the processing of the mail continues. If false it stops.
+ *  - "allowRepositoryCreation" optional, defaults to true. If true, non existing repository will be created. In case of
+ *  misconfiguration, this might lead to arbitrary repository creation. If false, the incoming mails will be stored only
+ *  in already existing repository. If not existing, the email will be dropped with an appropriate log warning (leading
+ *  to potential data loss).
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java]: Could we indicate to the admin how to create a mail repository he might need, and have a link to webadmin documentation?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    @Override
+    public void init() throws MessagingException {
+        urlPrefix = Optional.ofNullable(getInitParameter("urlPrefix"))
+            .orElseThrow(() -> new MessagingException("'urlPrefix' is a mandatory configuration property"));
+        passThrough = getInitParameter("passThrough", DEFAULT_CONSUME);
+    }
+
+    @Override
+    public void service(Mail mail) throws MessagingException {
+        String url = urlPrefix + mail.getSender().getDomain().asString();
```
Comment (mbaechler) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/ToSenderDomainRepository.java]: repositoryUrl ?

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+            .mailetName("Test")
+            .build();
+
+        assertThatThrownBy(() -> mailet.init(mailetConfig))
+            .isInstanceOf(MessagingException.class);
+    }
+
+    @Test
+    void initShouldNotThrowWhenUrlPrefixIsPresent() throws MessagingException {
+        mailet.init(DEFAULT_MAILET_CONFIG);
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: to make the intent obvious, you can use `assertThatCode().doesntThrowAnyException()`

File: server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+import org.apache.mailet.base.test.FakeMail;
+import org.apache.mailet.base.test.FakeMailetConfig;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class ToSenderDomainRepositoryTest {
+
+    private static final FakeMailetConfig DEFAULT_MAILET_CONFIG = FakeMailetConfig.builder()
+        .mailetName("TestConfig")
+        .setProperty("urlPrefix", "memory://var/mail/dlp/")
```
Comment (mbaechler) [inline server/mailet/mailets/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: you should put this url in a Constant, you copy/paste it everywhere

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+    public void tearDown() {
+        jamesServer.shutdown();
+    }
+
+    @Test
+    public void incomingMailShouldBeStoredInCorrespondingMailRepository() throws Exception {
+        messageSender.connect(LOCALHOST_IP, SMTP_PORT)
+            .sendMessage(RECIPIENT, RECIPIENT);
+
+        awaitAtMostOneMinute.until(
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: the await should not be the assertion, it's just a condition to go ahead in the test

File: server/container/guice/guice-common/src/main/java/org/apache/james/utils/DataProbeImpl.java [hunk 13]
```diff
@@ -58,13 +58,18 @@ public void addUser(String userName, String password) throws Exception {
         usersRepository.addUser(userName, password);
     }
 
+    public DataProbeImpl fluentAddUser(String userName, String password) throws Exception {
```
Comment (mbaechler) [inline server/container/guice/guice-common/src/main/java/org/apache/james/utils/DataProbeImpl.java]: why do you need to create a new method for that ? changing addUser to return `this` should be enough ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java [hunk 14]
```diff
@@ -54,49 +54,140 @@
...
             .sendMessage(RECIPIENT, RECIPIENT)
             .sendMessage(RECIPIENT, RECIPIENT);
 
         awaitAtMostOneMinute.until(
-            () -> probe.getRepositoryMailCount(CUSTOM_REPOSITORY_PREFIX + DEFAULT_DOMAIN) == 2);
+            () -> jamesServer.getProbe(MailRepositoryProbeImpl.class)
+                .getRepositoryMailCount(CUSTOM_REPOSITORY_PREFIX + DEFAULT_DOMAIN) == 2);
+    }
+
+    private void initializeJamesServer(MailetContainer.Builder mailetContainer) throws Exception {
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: this method looks weird to me. Usually, we use @Before to initialize things.
If you need a specific initialization process, you should name it to explain what is specific about it

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/MailRepositoryStoreService.java [hunk 15]
```diff
@@ -56,6 +56,9 @@ public MailRepositoryStoreService(MailRepositoryStore mailRepositoryStore) {
             .collect(Guavate.toImmutableList());
     }
 
+    public MailRepository createMailRepository(String repositoryUrl) throws MailRepositoryStore.MailRepositoryStoreException {
+        return mailRepositoryStore.select(repositoryUrl);
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/MailRepositoryStoreService.java]: really ? don't we have a proper `create` method ?

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 16]
```diff
@@ -106,13 +109,40 @@ public void setUp() throws Exception {
...
+        RestAssured.enableLoggingOfRequestAndResponseIfValidationFails();
     }
 
     @After
     public void tearDown() {
         webAdminServer.destroy();
     }
 
+    @Test
+    public void putMailRepositoryShouldReturnOkWhenRepositoryIsCreated() throws Exception {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: what if it already exists ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java (hunk 14 above)
Comment (trantienduchn) [reply server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: I can't place it inside @before method cause each test need different `MailetContainer.Builder`.  How about the name `startJamesServerWithMailetContainer`?

File: server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/MailRepositoryStoreService.java (hunk 15 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/service/MailRepositoryStoreService.java]: Nope, we should add it.

File: server/container/guice/guice-common/src/main/java/org/apache/james/utils/DataProbeImpl.java (hunk 13 above)
Comment (chibenwa) [reply server/container/guice/guice-common/src/main/java/org/apache/james/utils/DataProbeImpl.java]: The "big deal" is JMX silly stuff that expects serialization, thus we can not touch the DataProbe API

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java (hunk 12 above)
Comment (trantienduchn) [reply server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: I can't get it, can you provide an concrete example?

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java (hunk 14 above)
Comment (mbaechler) [reply server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: ok

File: server/container/guice/guice-common/src/main/java/org/apache/james/utils/DataProbeImpl.java (hunk 13 above)
Comment (mbaechler) [reply server/container/guice/guice-common/src/main/java/org/apache/james/utils/DataProbeImpl.java]: Ok, I see, could we add a  ̀.fluent() ̀ method that return a fluent API instead ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java (hunk 12 above)
Comment (mbaechler) [reply server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: yes.
await.until(X) is not the same as assertThat(Y).isSomething()
I mean, even if X is an assertion, the reader is not supposed to infer it's the expected result of the test.
You should write : await.until(x) then assertThat(Y).yadayada

File: server/data/data-api/src/main/java/org/apache/james/probe/DataProbe.java [hunk 17]
```diff
@@ -26,6 +26,33 @@
 
 public interface DataProbe {
 
+    class DataProbeFluent {
```
Comment (chibenwa) [inline server/data/data-api/src/main/java/org/apache/james/probe/DataProbe.java]: FluentDataProbe

**Comment by trantienduchn** [comment]
test this please

**Comment by trantienduchn** [comment]
test this please

File: server/data/data-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryStore.java [hunk 18]
```diff
@@ -33,6 +33,14 @@
      */
     MailRepository select(String url) throws MailRepositoryStoreException;
 
+    /**
+     * Create the {@link MailRepository} for the given url and return it self. If the repository is already existed,
```
Comment (rouazana) [inline server/data/data-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryStore.java]: s/it self/it ?
s/is already existed/already exists/ ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java [hunk 19]
```diff
@@ -0,0 +1,209 @@
...
+
+        awaitAtMostOneMinute.until(
+            () -> probe.getRepositoryMailCount(CUSTOM_REPOSITORY_PREFIX + DEFAULT_DOMAIN) == 1);
+
+        assertThat(probe.getRepositoryMailCount(CUSTOM_REPOSITORY_PREFIX + DEFAULT_DOMAIN))
+            .isEqualTo(1);
+    }
+
+    @Test
+    public void incomingMailShouldBeStoredWhenRepositoryDoNotExistAndAllowedToCreateRepository() throws Exception {
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: s/Do/Does

File: server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java [hunk 20]
```diff
@@ -0,0 +1,209 @@
...
+
+        awaitAtMostOneMinute.until(
+            () -> probe.getRepositoryMailCount(CUSTOM_REPOSITORY_PREFIX + DEFAULT_DOMAIN) == 1);
+
+        assertThat(probe.getRepositoryMailCount(CUSTOM_REPOSITORY_PREFIX + DEFAULT_DOMAIN))
+            .isEqualTo(1);
+    }
+
+    @Test
+    public void incomingMailShouldBeIgnoredWhenRepositoryDoNotExistAndNotAllowedToCreateRepository() throws Exception {
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/transport/mailets/ToSenderDomainRepositoryTest.java]: s/Do/Does

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 21]
```diff
@@ -106,13 +110,56 @@ public void setUp() throws Exception {
...
+            .put(URL_ESCAPED_MY_REPO)
+        .then()
+            .statusCode(HttpStatus.NO_CONTENT_204);
+
+        verify(mailRepositoryStore).create(URL_MY_REPO);
+        verifyNoMoreInteractions(mailRepositoryStore);
+    }
+
+    @Test
+    public void putMailRepositoryShouldReturnOkWhenRepositoryIsAlreadyExisted() throws Exception {
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: s/IsAlreadyExisted/AlreadyExists

File: server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java [hunk 22]
```diff
@@ -106,13 +110,56 @@ public void setUp() throws Exception {
...
+            .statusCode(HttpStatus.NO_CONTENT_204);
+
+        verify(mailRepositoryStore, times(2)).create(URL_MY_REPO);
+        verifyNoMoreInteractions(mailRepositoryStore);
+    }
+
+    @Test
+    public void putMailRepositoryShouldReturnServerErrorWhenCannotCreateRepository() throws Exception {
+        when(mailRepositoryStore.create(anyString()))
+            .thenThrow(new MailRepositoryStore.MailRepositoryStoreException("Error while select repository url://myRepo"));
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailrepository/src/test/java/org/apache/james/webadmin/routes/MailRepositoriesRoutesTest.java]: s/select/selecting

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
test this please (obiwan crashed)

**Comment by chibenwa** [comment]
```
[84bcfcda61490882c9087c5c860f91ce485559b6] Tests in error: 
[84bcfcda61490882c9087c5c860f91ce485559b6]   NetworkMatcherIntegrationTest.remoteAddrInNetworkShouldSupportRangesDefinedByAMiddleIp:165 » ConditionTimeout
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**